{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2023-06-23T15:11:36.904382Z","iopub.status.busy":"2023-06-23T15:11:36.903998Z","iopub.status.idle":"2023-06-23T15:11:36.920135Z","shell.execute_reply":"2023-06-23T15:11:36.918946Z","shell.execute_reply.started":"2023-06-23T15:11:36.904355Z"},"trusted":true},"outputs":[],"source":["import numpy as np\n","import random\n","import csv\n","from Bio import SeqIO\n","from tensorflow.keras.utils import plot_model, model_to_dot\n","import tensorflow as tf\n","from tensorflow.keras.metrics import AUC\n","from tensorflow.keras.models import Sequential\n","from tensorflow.keras.layers import Conv1D, MaxPooling1D, Dropout, LSTM, Dense, TimeDistributed, Bidirectional\n","from sklearn.model_selection import train_test_split\n","from tensorflow.keras.optimizers import Adam\n","from tensorflow.keras.metrics import Precision, Recall\n","from tensorflow.keras import backend as K\n","from tensorflow.keras.models import load_model\n","def f1_score(y_true, y_pred): \n","    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))\n","    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))\n","    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))\n","    precision = true_positives / (predicted_positives + K.epsilon())\n","    recall = true_positives / (possible_positives + K.epsilon())\n","    f1_val = 2*(precision*recall)/(precision+recall+K.epsilon())\n","    return f1_val\n","def negative_predictive_value(y_true, y_pred):\n","    true_negatives = K.sum(K.round(K.clip((1 - y_true) * (1 - y_pred), 0, 1)))\n","    total_negatives = K.sum(K.round(K.clip(1 - y_true, 0, 1)))\n","    return true_negatives / (total_negatives + K.epsilon())\n","\n","def one_hot_decode(encoded_seq):\n","    mapping = {tuple([1, 0, 0, 0]): 'A',\n","               tuple([0, 1, 0, 0]): 'C',\n","               tuple([0, 0, 1, 0]): 'G',\n","               tuple([0, 0, 0, 1]): 'T',\n","               tuple([0, 0, 0, 0]): 'x'}\n","    return ''.join(mapping[tuple(nt)] for nt in encoded_seq)\n","\n","def one_hot_encode(sequence):\n","    mapping = {'A': [1, 0, 0, 0], 'C': [0, 1, 0, 0], 'G': [0, 0, 1, 0], 'T': [0, 0, 0, 1]}\n","    return np.array([mapping[nt] for nt in sequence])\n"]},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","execution":{"iopub.execute_input":"2023-06-23T15:11:40.579635Z","iopub.status.busy":"2023-06-23T15:11:40.579278Z","iopub.status.idle":"2023-06-23T15:11:45.454593Z","shell.execute_reply":"2023-06-23T15:11:45.453574Z","shell.execute_reply.started":"2023-06-23T15:11:40.579606Z"},"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["4494\n"]}],"source":["\n","# učitavanje podataka za treniranje\n","\n","gff_file = '/path'\n","fna_file = '/path'\n","\n","genes = []\n","with open(gff_file, 'r') as f:\n","    reader = csv.reader(f, delimiter='\\t')\n","    for row in reader:\n","        if row and not row[0].startswith('#') and row[2] == 'gene':\n","            start = int(row[3]) - 1\n","            end = int(row[4])\n","            genes.append(( start, end))\n","\n","genome_array = []\n","label_array = []\n","for record in SeqIO.parse(fna_file, 'fasta'):\n","    genome_array = one_hot_encode(str(record.seq))\n","    label_array = np.zeros(len(record.seq))\n","    for  start, end in genes:\n","        label_array[start:end] = 1\n","block_size = 500\n","step_size = 250  # Step size for the sliding window\n","genome_blocks = [genome_array[i:i+block_size] for i in range(0, len(genome_array) - block_size + 1, step_size)]\n","label_blocks = [label_array[i:i+block_size] for i in range(0, len(label_array) - block_size + 1, step_size)]\n","\n","\n","data = list(zip(X, y))\n","random.shuffle(data)\n","X, y = zip(*data)\n","\n","X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n","X_test,y_test= X,y\n","X_train = np.array(X_train)\n","X_test = np.array(X_test)\n","y_train = np.array(y_train)[:,:,None]\n","y_test = np.array(y_test)[:,:,None]\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"execution":{"iopub.execute_input":"2023-06-12T19:16:31.314793Z","iopub.status.busy":"2023-06-12T19:16:31.314097Z","iopub.status.idle":"2023-06-12T19:51:01.399991Z","shell.execute_reply":"2023-06-12T19:51:01.398330Z","shell.execute_reply.started":"2023-06-12T19:16:31.314761Z"},"jupyter":{"outputs_hidden":true,"source_hidden":true},"trusted":true},"outputs":[],"source":["model = Sequential([\n","    Conv1D(filters=32, kernel_size=9, activation='relu', padding='same', input_shape=(500, 4)),\n","    Dropout(0.2),\n","    Conv1D(filters=64, kernel_size=9, activation='relu', padding='same'),\n","    Dropout(0.2),\n","    Bidirectional(LSTM(256, return_sequences=True)),\n","    Bidirectional(LSTM(256, return_sequences=True)),\n","    TimeDistributed(Dense(256, activation='relu')),\n","    TimeDistributed(Dense(128, activation='relu')),\n","    TimeDistributed(Dense(1, activation='sigmoid'))\n","])\n","\n","optimizer = Adam(learning_rate=0.001)\n","model.compile(optimizer=\"adam\",\n","              loss='binary_crossentropy',\n","              metrics=['accuracy',negative_predictive_value, Precision(name='precision'), Recall(name='recall'), f1_score])\n","\n","model.fit(X_train, y_train, epochs=50, batch_size=16, validation_split=0.2)\n","#model.save('')\n","loss, accuracy,Npv,precision,recall,f1_score = model.evaluate(X_test, y_test)\n","print(f'Test accuracy: {accuracy:.2f}')\n","print(loss)\n"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2023-06-23T15:11:45.457022Z","iopub.status.busy":"2023-06-23T15:11:45.456458Z","iopub.status.idle":"2023-06-23T15:11:58.667260Z","shell.execute_reply":"2023-06-23T15:11:58.666242Z","shell.execute_reply.started":"2023-06-23T15:11:45.456989Z"},"trusted":true},"outputs":[],"source":["\n","model=load_model(\"/path\")"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["#učitavanje podataka za predviđanje \n","gff_file = '/path'\n","fna_file = '/path'\n","\n","genes = []\n","with open(gff_file, 'r') as f:\n","    reader = csv.reader(f, delimiter='\\t')\n","    for row in reader:\n","        if row and not row[0].startswith('#') and row[2] == 'gene':\n","            start = int(row[3]) - 1\n","            end = int(row[4])\n","            genes.append(( start, end))\n","\n","genome_array = []\n","label_array = []\n","for record in SeqIO.parse(fna_file, 'fasta'):\n","    genome_array = one_hot_encode(str(record.seq))\n","    label_array = np.zeros(len(record.seq))\n","    for  start, end in genes:\n","        label_array[start:end] = 1\n","block_size = 500\n","genome_blocks = [genome_array[i:i+block_size] for i in range(0, len(genome_array), block_size)]\n","label_blocks = [label_array[i:i+block_size] for i in range(0, len(label_array), block_size)]\n","\n","genome_blocks = [block for block in genome_blocks if len(block) == block_size]\n","label_blocks = [block for block in label_blocks if len(block) == block_size]\n","\n","X = np.array(genome_blocks)\n","y = np.array(label_blocks)\n","\n","X_test,y_test= X,y\n","X_test = np.array(X_test)\n","y_test = np.array(y_test)[:,:,None]\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2023-06-23T15:12:02.069530Z","iopub.status.busy":"2023-06-23T15:12:02.069125Z","iopub.status.idle":"2023-06-23T15:12:44.939636Z","shell.execute_reply":"2023-06-23T15:12:44.938535Z","shell.execute_reply.started":"2023-06-23T15:12:02.069502Z"},"trusted":true},"outputs":[],"source":["\n","predictions = model.predict(X_test)\n","binary_predictions = np.where(predictions > 0.45, 1, 0)"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2023-06-23T15:13:50.873630Z","iopub.status.busy":"2023-06-23T15:13:50.873270Z","iopub.status.idle":"2023-06-23T15:14:05.702660Z","shell.execute_reply":"2023-06-23T15:14:05.701591Z","shell.execute_reply.started":"2023-06-23T15:13:50.873600Z"},"trusted":true},"outputs":[],"source":["#ispis sekvence na kojoj je označeno što je kodirajući a sto nekodirajući dio sekvence\n","predseq=[]\n","for j in range(0,len(binary_predictions)):\n","    test1=binary_predictions[j]\n","    for i in range(0,len(test1)):\n","        #print(test1[i])\n","        if test1[i][0]==1:\n","            predseq.append(X_test[j][i])\n","        else:\n","            predseq.append(tuple([0, 0, 0, 0]))\n","        \n","print(len(predseq))\n","sequence = one_hot_decode(predseq)\n","#print(sequence)\n","    "]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2023-06-23T15:14:10.685538Z","iopub.status.busy":"2023-06-23T15:14:10.685179Z","iopub.status.idle":"2023-06-23T15:14:10.747647Z","shell.execute_reply":"2023-06-23T15:14:10.746578Z","shell.execute_reply.started":"2023-06-23T15:14:10.685509Z"},"trusted":true},"outputs":[],"source":["# brojanje gena i ispis dobivenih gena \n","def count_coding_sequences(dna_seq):\n","    parts = [s for s in dna_seq.split('X') if len(s)>50]\n","    for i in range(10):  \n","        print(parts[i],\"\\n\")\n","    count = len(parts)\n","\n","    return count\n","print(count_coding_sequences(sequence))"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2023-06-23T15:14:20.032068Z","iopub.status.busy":"2023-06-23T15:14:20.031690Z","iopub.status.idle":"2023-06-23T15:14:45.531261Z","shell.execute_reply":"2023-06-23T15:14:45.530248Z","shell.execute_reply.started":"2023-06-23T15:14:20.032038Z"},"trusted":true},"outputs":[],"source":["# isti proces kao u čelijama gore samo za stvarne gene\n","predseq=[]\n","for j in range(0,len(y_test)):\n","    test2=y_test[j]\n","    for i in range(0,len(test2)):\n","        if test2[i]==1:\n","            predseq.append(X_test[j][i])\n","        else:\n","            predseq.append(tuple([0, 0, 0, 0]))\n","        \n","print(len(predseq))\n","sequence = one_hot_decode(predseq)\n","#print(sequence)"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2023-06-23T15:14:50.365501Z","iopub.status.busy":"2023-06-23T15:14:50.365126Z","iopub.status.idle":"2023-06-23T15:14:50.413226Z","shell.execute_reply":"2023-06-23T15:14:50.411924Z","shell.execute_reply.started":"2023-06-23T15:14:50.365471Z"},"trusted":true},"outputs":[],"source":["def count_coding_sequences(dna_seq):\n","    parts = [s for s in dna_seq.split('X') if s]\n","    for i in range(10): \n","        print(parts[i],\"\\n\")\n","\n","    count = len(parts)\n","\n","    return count\n","print(count_coding_sequences(sequence))"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2023-06-20T20:38:43.756469Z","iopub.status.busy":"2023-06-20T20:38:43.756080Z","iopub.status.idle":"2023-06-20T20:40:36.700868Z","shell.execute_reply":"2023-06-20T20:40:36.699689Z","shell.execute_reply.started":"2023-06-20T20:38:43.756436Z"},"trusted":true},"outputs":[],"source":["test_fna_file = 'path'\n","test_gff_file = 'path'\n","\n","test_genes = []\n","with open(test_gff_file, 'r') as f:\n","    reader = csv.reader(f, delimiter='\\t')\n","    for row in reader:\n","        if row and not row[0].startswith('#') and row[2] == 'gene':\n","            start = int(row[3]) - 1\n","            end = int(row[4])\n","            strand = 1 if row[6] == '+' else -1\n","            test_genes.append((start, end))\n","\n","test_genome_array = []\n","test_label_array = []\n","for record in SeqIO.parse(test_fna_file, 'fasta'):\n","    test_genome_array = one_hot_encode(str(record.seq))\n","    test_label_array = np.zeros(len(record.seq))\n","    for start, end in test_genes:\n","        test_label_array[start:end] = 1\n","\n","block_size = 500\n","step_size = 250\n","\n","test_genome_blocks = [test_genome_array[i:i+block_size] for i in range(0, len(test_genome_array) - block_size + 1, step_size)]\n","test_label_blocks = [test_label_array[i:i+block_size] for i in range(0, len(test_label_array) - block_size + 1, step_size)]\n","\n","test_genome_blocks = [block for block in test_genome_blocks if len(block) == block_size]\n","test_label_blocks = [block for block in test_label_blocks if len(block) == block_size]\n","\n","X_test = np.array(test_genome_blocks)\n","y_test = np.array(test_label_blocks)\n","\n","y_test = y_test[:,:,None]\n","predictions = model.predict(X_test)\n","final_prediction = np.zeros(len(test_genome_array))\n","\n","for i, block in enumerate(predictions):\n","    start_index = i * step_size\n","    end_index = start_index + block_size\n","    final_prediction[start_index:end_index] += block.flatten()\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"collapsed":true,"execution":{"iopub.execute_input":"2023-06-20T21:32:29.827316Z","iopub.status.busy":"2023-06-20T21:32:29.826917Z","iopub.status.idle":"2023-06-20T21:32:43.941237Z","shell.execute_reply":"2023-06-20T21:32:43.939775Z","shell.execute_reply.started":"2023-06-20T21:32:29.827287Z"},"jupyter":{"outputs_hidden":true},"trusted":true},"outputs":[],"source":["#ispis predikcija\n","binary_predictions = np.where(final_prediction > 1, 1, 0)\n","for i in range(len(binary_predictions)):\n","    print(binary_predictions[i],end = \"\")\n","    #print(int(y_test.flatten()[i]))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["#ispis predikcije u obliku sekvence genoma\n","predseq=[]\n","for j in range(0,len(binary_predictions[:6000]),block_size):\n","    test1=binary_predictions[j:j+block_size]\n","    for i in range(0,len(test1)):\n","        if test1[i]==1:\n","            predseq.append(X_test[j][i])\n","        else:\n","            predseq.append(tuple([0, 0, 0, 0]))\n","        \n","print(len(predseq))\n","sequence = one_hot_decode(predseq)\n","print(sequence)"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["\n"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.10"}},"nbformat":4,"nbformat_minor":4}
